File: java/grade-school/src/main/java/School.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class School {

    private Map<String, Integer> roster;
    public School() {
        this.roster = new HashMap<>();
    }

    boolean add(String student, int grade) {
        if(this.roster.containsKey(student)) {
            return false;
        }

        this.roster.put(student, grade);
        return true;
    }

    List<String> roster() {
        return this.roster.entrySet().stream()
                    .sorted((a, b) -> { 
                        int res = a.getValue() - b.getValue();
                        return res == 0 ? a.getKey().compareTo(b.getKey()) : res;
                    })
                    .map(s -> s.getKey()).toList();
    }

    List<String> grade(int grade) {
        return this.roster.entrySet().stream()
                .filter(student -> student.getValue() == grade)
                .map(student -> student.getKey())
                .sorted(String::compareTo).toList();
    }

}

File: elixir/grade-school/lib/school.ex
defmodule School do
  @moduledoc """
  Simulate students in a school.

  Each student is in a grade.
  """

  @type school :: map()

  @doc """
  Create a new, empty school.
  """
  @spec new() :: school
  def new() do
    Map.new()
  end

  @doc """
  Add a student to a particular grade in school.
  """
  @spec add(school, String.t(), integer) :: {:ok | :error, school}
  def add(school, name, grade) do
    current_roster = roster(school)

    if name in current_roster do
      {:error, school}
    else
      students = Map.get(school, grade, [])
      {:ok, Map.put(school, grade, students ++ [name])}
    end
  end

  @doc """
  Return the names of the students in a particular grade, sorted alphabetically.
  """
  @spec grade(school, integer) :: [String.t()]
  def grade(school, grade) do
    Map.get(school, grade, [])
    |> Enum.sort()
  end

  @doc """
  Return the names of all the students in the school sorted by grade and name.
  """
  @spec roster(school) :: [String.t()]
  def roster(school) do
    Map.keys(school)
    |> Enum.sort()
    |> Enum.reduce([], fn key, students ->
      students ++ grade(school, key)
    end)
  end
end
